Add support for $value_type params to get_option(), get_metadata(), etc.

azaozz opened the pull request:
Trac ticket: https://core.trac.wordpress.org/ticket/55942

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

azaozz committed:
Add $value_type params

azaozz committed:
Merge remote-tracking branch 'upstream/trunk' into trunk

azaozz committed:
Fix one empty space..

petitphp commented on src/wp-includes/functions.php:
```diff
@@ -8509,3 +8509,69 @@ function is_php_version_compatible( $required ) {
+ * @return mixed The decoded value.
+ */
+function wp_decode_value_from_db( $value, $value_type = '' ) {
+	$origianl_value = $value;
```
Small typo here `$origianl_value`, also present in the `apply_filters`

azaozz replied:
Fixed, thanks for spotting it :)

petitphp reviewed (commented):
This is a more simpler and straightforward approach than what I've worked on in #3124 which is more convoluted.

I like the `$value_type` parameter and the fact that just reading the code it allows to quickly see what the type of a meta or an option with consistency.

One thing I've noted, the `get_{post|term|user}_meta` functions have not been updated to support the new `$value_type` parameter. Is it something you'll be adding ?

azaozz committed:
Merge remote-tracking branch 'upstream/trunk' into trunk

azaozz committed:
Fix typo

azaozz committed:
Add $value_type params to all get_*_meta functions

azaozz commented:
>This is a more simpler and straightforward approach

Yea, unfortunately it seems very complex to implement the original idea and keep 100% back-compat. So this PR is a "cut-down" version without any DB changes that still brings some of the original benefits. 

> One thing I've noted, the get_{post|term|user}_meta functions have not been updated

Updated them. Next would be to add more tests and also some existing `get_{post|term|user|comment}_meta` tests to include the new param.

azaozz committed:
Add @since

azaozz committed:
Update inline docs

azaozz committed:
Merge remote-tracking branch 'upstream/trunk' into trunk

azaozz commented:
Been wondering if it is a good idea to enforce the `$value_type` to the values returned from the "pre" filters: `pre_option_{$option}`, `pre_option`, `get_{$meta_type}_metadata`. 

Seems the type should be enforced when used as the function that is getting the option or the meta is specifying the value type. Perhaps that may bring some unexpected results sometimes. However not enforcing the type also may bring unexpected results.

Any ideas or thoughts about this? :)

azaozz committed:
Ensure proper value types are returned from the "pre" filters

Also introduce `wp_settype_to_value_from_db()`.

azaozz committed:
Merge remote-tracking branch 'upstream/trunk' into trunk

azaozz committed:
Fix returning a non-array from the pre_meta filter for multiple meta.

petitphp commented:
> Been wondering if it is a good idea to enforce the `$value_type` to the values returned from the "pre" filters: `pre_option_{$option}`, `pre_option`, `get_{$meta_type}_metadata`.
> 
> Seems the type should be enforced when used as the function that is getting the option or the meta is specifying the >value type. Perhaps that may bring some unexpected results sometimes. However not enforcing the type also may bring >unexpected results.

I agree with this. When I specify a value type the function should try to always return the expected type even if the value come from a filter.

azaozz committed:
More docs,

ensure expected type is returned from the `get_{$meta_type}_metadata_by_mid` filter.

azaozz committed:
Fix typo.

azaozz committed:
Merge remote-tracking branch 'upstream/trunk' into trunk

azaozz committed:
Merge remote-tracking branch 'upstream/trunk' into trunk

azaozz committed:
Add 'int' and 'bool' to the list of possible value types

Ensure get_network_option() respects $value_type when set.

ironprogrammer committed:
Move tests into dedicated class w/ data provider

ironprogrammer committed:
Remove migrated tests from functions test class

azaozz committed:
Merge pull request #1 from ironprogrammer/update-tests

Update tests. Thanks @ironprogrammer!

azaozz committed:
Fix (auto) whitespaces.

azaozz committed:
Fix required unsightly array formatting

azaozz committed:
Merge remote-tracking branch 'upstream/trunk' into trunk

azaozz committed:
Merge remote-tracking branch 'upstream/trunk' into trunk

costdev commented on src/wp-includes/functions.php:
```diff
@@ -8539,3 +8539,93 @@ function is_php_version_compatible( $required ) {
 }
+
+/**
+ * Get supported data types for meta and option values.
```
```suggestion
 * Gets supported data types for meta and option values.
```

costdev commented on src/wp-includes/functions.php:
```diff
@@ -8539,3 +8539,93 @@ function is_php_version_compatible( $required ) {
+}
+
+/**
+ * Set the type of a meta or option value that was retrieved from the database.
```
```suggestion
 * Sets the type of a meta or option value that was retrieved from the database.
```

costdev commented on src/wp-includes/functions.php:
```diff
@@ -8539,3 +8539,93 @@ function is_php_version_compatible( $required ) {
+			settype( $value, $value_type );
+		} else {
+			// Improper use.
+			/* translators: The function name that returns the supported types. */
+			$message = sprintf( __( 'Only types returned by %s are supported.' ), '<code>wp_get_db_value_types()</code>' );
+			_doing_it_wrong( 'wp_settype_to_value_from_db', $message );
```
```suggestion
			
			_doing_it_wrong(
				'wp_settype_to_value_from_db',
				sprintf(
					/* translators: %s: The function name that returns the supported types. */
					__( 'Only types returned by %s are supported.' ),
					'<code>wp_get_db_value_types()</code>'
				)
			);
```
1. We don't need to store the message in a variable.
2. By passing the `sprintf()` directly to `_doing_it_wrong()`, we can have the `// Improper use.` inline comment separated from the `/* translators */` comment.
3. Added `%s:` to the `/*translators */` comment.

costdev commented on src/wp-includes/functions.php:
```diff
@@ -8539,3 +8539,93 @@ function is_php_version_compatible( $required ) {
+
+
+/**
+ * Decode a value after it was retrieved from the database.
```
```suggestion
 * Decodes a value after it was retrieved from the database.
```

costdev commented on src/wp-includes/functions.php:
```diff
@@ -8539,3 +8539,93 @@ function is_php_version_compatible( $required ) {
+	}
+
+	/**
+	 * Filter the decoded value.
```
```suggestion
	 * Filters the decoded value.
```

costdev commented on src/wp-includes/meta.php:
```diff
@@ -550,26 +550,32 @@ function delete_metadata( $meta_type, $object_id, $meta_key, $meta_value = '', $
+ *                           the specified object. Default empty string.
+ * @param bool   $single     Optional. If true, return only the first value of the specified `$meta_key`.
+ *                           This parameter has no effect if `$meta_key` is not specified. Default false.
+ * @param string $value_type Optional. The expected data type of the value.
```
```suggestion
 * @param string $value_type Optional. The expected data type of the value. Default empty string.
```

costdev commented on src/wp-includes/meta.php:
```diff
@@ -581,21 +587,23 @@ function get_metadata( $meta_type, $object_id, $meta_key = '', $single = false )
+ *                           the specified object. Default empty string.
+ * @param bool   $single     Optional. If true, return only the first value of the specified `$meta_key`.
+ *                           This parameter has no effect if `$meta_key` is not specified. Default false.
+ * @param string $value_type Optional. The expected data type of the value.
```
```suggestion
 * @param string $value_type Optional. The expected data type of the value. Default empty string.
```

costdev commented on src/wp-includes/meta.php:
```diff
@@ -766,12 +805,14 @@ function metadata_exists( $meta_type, $object_id, $meta_key ) {
+ * @param string $meta_type  Type of object metadata is for. Accepts 'post', 'comment', 'term', 'user',
+ *                           or any other object type with an associated meta table.
+ * @param int    $meta_id    ID for a specific meta row.
+ * @param string $value_type Optional. The expected data type of the value.
```
```suggestion
 * @param string $value_type Optional. The expected data type of the value. Default empty string.
```

costdev commented on src/wp-includes/ms-site.php:
```diff
@@ -1061,20 +1061,22 @@ function delete_site_meta( $site_id, $meta_key, $meta_value = '' ) {
+ * @param bool   $single     Optional. Whether to return a single value.
+ *                           This parameter has no effect if `$key` is not specified.
+ *                           Default false.
+ * @param string $value_type Optional. The expected data type of the value.
```
```suggestion
 * @param string $value_type Optional. The expected data type of the value. Default empty string.
```

costdev commented on src/wp-includes/option.php:
```diff
@@ -64,18 +72,20 @@
  *
  * @param string $option        Name of the option to retrieve. Expected to not be SQL-escaped.
  * @param mixed  $default_value Optional. Default value to return if the option does not exist.
+ * @param string $value_type    Optional. The expected data type of the value.
```
```suggestion
 * @param string $value_type    Optional. The expected data type of the value. Default empty string.
```

costdev commented on src/wp-includes/option.php:
```diff
@@ -1412,9 +1430,10 @@ function update_site_option( $option, $value ) {
  * @param int    $network_id    ID of the network. Can be null to default to the current network ID.
  * @param string $option        Name of the option to retrieve. Expected to not be SQL-escaped.
  * @param mixed  $default_value Optional. Value to return if the option doesn't exist. Default false.
+ * @param string $value_type    Optional. The expected data type of the value.
```
```suggestion
 * @param string $value_type    Optional. The expected data type of the value. Default empty string.
```

costdev commented on src/wp-includes/post.php:
```diff
@@ -2488,20 +2488,22 @@ function delete_post_meta( $post_id, $meta_key, $meta_value = '' ) {
+ * @param bool   $single     Optional. Whether to return a single value.
+ *                           This parameter has no effect if `$key` is not specified.
+ *                           Default false.
+ * @param string $value_type Optional. The expected data type of the value.
```
```suggestion
 * @param string $value_type Optional. The expected data type of the value. Default empty string.
```

costdev commented on src/wp-includes/taxonomy.php:
```diff
@@ -1371,20 +1371,22 @@ function delete_term_meta( $term_id, $meta_key, $meta_value = '' ) {
+ * @param bool   $single     Optional. Whether to return a single value.
+ *                           This parameter has no effect if `$key` is not specified.
+ *                           Default false.
+ * @param string $value_type Optional. The expected data type of the value.
```
```suggestion
 * @param string $value_type Optional. The expected data type of the value. Default empty string.
```

costdev commented on src/wp-includes/user.php:
```diff
@@ -1152,22 +1152,24 @@ function delete_user_meta( $user_id, $meta_key, $meta_value = '' ) {
+ * @param bool   $single     Optional. Whether to return a single value.
+ *                           This parameter has no effect if `$key` is not specified.
+ *                           Default false.
+ * @param string $value_type Optional. The expected data type of the value.
```
```suggestion
 * @param string $value_type Optional. The expected data type of the value. Default empty string.
```

costdev commented on tests/phpunit/tests/functions/wpDecodeValueFromDb.php:
```diff
@@ -0,0 +1,91 @@
+ * @group functions.php
+ * @group meta
+ * @group option
+ *
+ * @ticket 55942
```
```suggestion
```
Test classes typically don't use `@ticket`, which is used for test methods, as additional test methods would not be added by this ticket number.

costdev commented on tests/phpunit/tests/functions/wpDecodeValueFromDb.php:
```diff
@@ -0,0 +1,91 @@
+
+	/**
+	 * Tests that wp_decode_value_from_db() returns the correct value.
+	 *
```
Each test method will need an `@ticket` annotation:

```php
/**
 * Tests that...
 *
 * @ticket
 *
 * @dataProvider
 *...
```

costdev commented on tests/phpunit/tests/functions/wpDecodeValueFromDb.php:
```diff
@@ -0,0 +1,91 @@
+	 * @return array
+	 */
+	public function data_wp_decode_value_from_db() {
+		$boolean = array(
```
If `bool` and `int` are also valid values, these should be tested as well.

costdev commented on tests/phpunit/tests/functions/wpDecodeValueFromDb.php:
```diff
@@ -0,0 +1,91 @@
+	/**
+	 * Tests that wp_decode_value_from_db() returns the correct value for objects.
+	 */
+	public function test_object_wp_decode_value_from_db() {
```
As this tests different types of objects, this should use a data provider with a dataset for each type.

costdev reviewed (changes requested):
Thanks for the PR @azaozz! I've left some feedback below.

azaozz committed:
Merge remote-tracking branch 'upstream/trunk' into trunk

azaozz committed:
Merge remote-tracking branch 'upstream/trunk' into trunk

costdev commented on src/wp-includes/functions.php:
```diff
@@ -8582,3 +8582,93 @@ function is_php_version_compatible( $required ) {
+ *
+ * @return array List of supported value types.
+ */
+function wp_get_db_value_types() {
```
Is it worth having a separate function for this array, rather than just defining it in `wp_settype_to_value_from_db()`? What are the other use cases you have in mind?

costdev commented on src/wp-includes/functions.php:
```diff
@@ -8582,3 +8582,93 @@ function is_php_version_compatible( $required ) {
+ * @param string $value_type The expected value type.
+ * @return mixed The value with the requested type.
+ */
+function wp_settype_to_value_from_db( $value, $value_type ) {
```
This name reads a little off - and `_settype_` is maybe too coupled to implementation. 

What about one of these?
- `wp_set_db_value_to_type()`
- `wp_set_value_from_db_to_type()`
- `wp_convert_db_value()`
- `wp_convert_db_value_to_type()`
- `wp_convert_value_from_db()`
- `wp_set_db_value_type()`

costdev commented on src/wp-includes/functions.php:
```diff
@@ -8582,3 +8582,93 @@ function is_php_version_compatible( $required ) {
+ * @since 6.3.0
+ *
+ * @param mixed  $value      Raw value from the database.
+ * @param string $value_type Optional. The expected value type.
```
```suggestion
 * @param string $value_type Optional. The expected value type. Default empty string.
```

costdev commented on src/wp-includes/functions.php:
```diff
@@ -8582,3 +8582,93 @@ function is_php_version_compatible( $required ) {
+	}
+
+	/**
+	 * Filter the decoded value.
```
```suggestion
	 * Filters the decoded value.
```

azaozz committed:
Merge remote-tracking branch 'upstream/trunk' into trunk

azaozz commented:
The trac ticket was closed.
